src/Components/Web/src/Virtualization/VirtualizeDeferred.cs

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. 
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+
+namespace Microsoft.AspNetCore.Components.Virtualization
+{
+    /// <summary>
+    /// Provides functionality for rendering a virtualized list from a deferred source.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public sealed class VirtualizeDeferred<TItem> : VirtualizeBase<TItem>
+    {
+        private enum FetchState
+        {
+            Idle,
+            Waiting,
+            Busy
+        }
+
+        private readonly ConcurrentQueue<TItem> _loadedItems = new ConcurrentQueue<TItem>();
+
+        private FetchState _fetchState;
+
+        private int _itemCount;
+
+        private Task<ItemsProviderResult<TItem>>? _fetchTask;
+
+        private CancellationTokenSource? _fetchTcs;
```

Comment:
Minor detail - should this be renamed to `_fetchCts`? I find it hard to remember which way round we go so I might be wrong, but it sounds more like an acronym for `CancellationTokenSource`.

Reply:
It should be - I was using `TaskCompletionSource` recently and the muscle memory from that acronym must have taken over. Oops!